reasoner/planner/CostEstimator.java

```diff
@@ -0,0 +1,384 @@
+            // Small inaccuracy: We double count duplicate roles (r:$a, r:$b)
+            // counts the case where r:$a=r:$b, which TypeDB wouldn't return
+            Set<RelationConstraint.RolePlayer> rolePlayers = concludable.relation().players();
+            List<Variable> constrainedVars = new ArrayList<>();
+            Map<TypeVariable, Long> rolePlayerEstimates = new HashMap<>();
+            Map<TypeVariable, Integer> rolePlayerCounts = new HashMap<>();
+
+            double relationTypeEstimate = countPersistedThingsMatchingType(concludable.relation().owner());
+            constrainedVars.add(concludable.relation().owner());
+            for (RelationConstraint.RolePlayer rp : rolePlayers) {
+                constrainedVars.add(rp.player());
+                TypeVariable key = rp.roleType().orElse(null);
+                rolePlayerCounts.put(key, rolePlayerCounts.getOrDefault(key, 0) + 1);
+                if (!rolePlayerEstimates.containsKey(key)){
+                    rolePlayerEstimates.put(key, countPersistedRolePlayers(rp));
+                }
+            }
+
+            return new LocalEstimate.CoPlayerEstimate(concludable, constrainedVars, relationTypeEstimate, rolePlayerEstimates, rolePlayerCounts);
+
+        }
+
+        private long countPersistedRolePlayers(RelationConstraint.RolePlayer rolePlayer) {
+            return logicMgr.graph().data().stats().thingVertexSum(rolePlayer.inferredRoleTypes());
+        }
+
+        private long countPersistedThingsMatchingType(ThingVariable thingVar) {
+            return logicMgr.graph().data().stats().thingVertexSum(thingVar.inferredTypes());
+        }
+
+        private long countPersistedHasEdges(Set<Label> ownerTypes, Set<Label> attributeTypes) {
+            return ownerTypes.stream().map(ownerType ->
+                    attributeTypes.stream()
+                            .map(attrType -> logicMgr.graph().data().stats().hasEdgeCount(ownerType, attrType))
+                            .reduce(0L, Long::sum)
+            ).reduce(0L, Long::sum);
+        }
+
+        private static abstract class LocalEstimate {
+
+            final List<Variable> variables;
+
+            private LocalEstimate(List<Variable> variables) { this.variables = variables; }
+
+            abstract long answerEstimate(ConjunctionAnswerCountEstimator costEstimator, Set<Variable> variablesOfInterest);
+
+            private static class SimpleEstimate extends LocalEstimate {
+                private final long staticEstimate;
+
+                private SimpleEstimate(List<Variable> variables, long estimate) {
+                    super(variables);
+                    this.staticEstimate = estimate;
+                }
+
+                @Override
+                long answerEstimate(ConjunctionAnswerCountEstimator costEstimator, Set<Variable> variablesOfInterest) { return staticEstimate; }
+            }
+
+            public static class CoPlayerEstimate extends LocalEstimate {
+
```

Comment:
(Possibly overkill)
Variables which play a role in the same relation are strongly co-constrained by the number of relations.
e.g. : The number of edges in a graph v/s the square of number of vertices. 

Considers the fact that multiple entities can play the same role in the same relation.
Supports answering queries on subsets of the role-players.